### PR TITLE
CodeQL query to detect XSLT injections

### PR DESCRIPTION
**Pull request by ggolawski**
This PR adds a query to detect XSLT injections. It flags the code where user-provided XSLT stylesheet is processed by `Transformer.transform`. The following use cases are supported:

* XSLT stylesheet provided as `StreamSource`:
```java
StreamSource source = new StreamSource(socket.getInputStream());
TransformerFactory.newInstance().newTransformer(source).transform();
```

* XSLT stylesheet provided as `SAXSource`:
```java
SAXSource source = new SAXSource(new InputSource(socket.getInputStream()));
TransformerFactory.newInstance().newTransformer(source).transform();
```

* XSLT stylesheet provided as `StAXSource`:
```java
StAXSource source = new StAXSource(XMLInputFactory.newInstance().createXMLEventReader(socket.getInputStream()));
TransformerFactory.newInstance().newTransformer(source).transform();
```

* XSLT stylesheet provided as `DOMSource`:
```java
DOMSource source = new DOMSource(DocumentBuilderFactory.newInstance().newDocumentBuilder().parse(socket.getInputStream()));
TransformerFactory.newInstance().newTransformer(source).transform();
```

* `Transformer` object created from `Templates`:
```java
TransformerFactory.newInstance().newTemplates(source).newTransformer().transform();
```

Processing of unvalidated XSLT stylesheets can lead to XXE or remote code execution.

This query partially overlaps with XXE query from `XXE.ql`, but has the following noticeable differences:

* XXE query doesn't raise the flag if `ACCESS_EXTERNAL_STYLESHEET` and `ACCESS_EXTERNAL_SCHEMA` are disabled. Disabling these options is enough to prevent XXE, but not enough to prevent RCE via XSLT injection. To prevent it, `FEATURE_SECURE_PROCESSING` must be enabled.
* In some cases, XXE query highlights the source code line where XSLT stylesheet is parsed (e.g. to create a `Document`). XsltInjection query always highlights the source code line where the transformation happens (`Transformer.transform` method invocation) - this is the place where XSLT injection (which can lead to RCE) happens.
* The case where `Transformer` is created from `Templates` (`TransformerFactory.newTemplates(source).newTransformer().transform()`) is not supported by XXE query.

The tests are also included.

### CODE REVIEW

**Comment by pwntester** [comment]
@ggolawski could you please add support for SAXON (e.g: `net.sf.saxon.s9api.XsltCompiler`) and XALAN (e.g: `org.apache.xalan.xslt.XSLTProcessor`)?

**Comment by ggolawski** [comment]
@pwntester I'll add support for SAXON.
Regarding Xalan, it looks like `org.apache.xalan.xslt.XSLTProcessor` was part of Xalan 1.x released in 2000 (?). It's very hard to find any distribution or documentation. It doesn't exist in Maven central either. And there're only ~200 references to `org.apache.xalan.xslt.XSLTProcessor` on GitHub (https://github.com/search?l=Java&q=%22org.apache.xalan.xslt.XSLTProcessor%22&type=Code).
Xalan 2.x implements `javax.xml.transform` interface and is handled by the current query implementation.

Does it make sense to add support for Xalan 1.x? What do you think?

**Comment by pwntester** [comment]
Makes sense not to support 1.x than, thanks!

**Comment by ggolawski** [comment]
@pwntester I've addedd support for Saxon. The following cases are supported:

```java
public void testSaxon(Socket socket) throws Exception {
    StreamSource source = new StreamSource(socket.getInputStream());
    XsltCompiler compiler = new Processor(true).newXsltCompiler();
    
    compiler.compile(source).load().transform();
    compiler.compile(source).load30().transform(null, null);
    compiler.compile(source).load30().applyTemplates((Source) null);
    compiler.compile(source).load30().applyTemplates((Source) null, null);
    compiler.compile(source).load30().applyTemplates((XdmValue) null);
    compiler.compile(source).load30().applyTemplates((XdmValue) null, null);
    compiler.compile(source).load30().callFunction(null, null);
    compiler.compile(source).load30().callFunction(null, null, null);
    compiler.compile(source).load30().callTemplate(null);
    compiler.compile(source).load30().callTemplate(null, null);
  }

  public void testSaxonXsltPackage(@RequestParam String param, Socket socket) throws Exception {
    URI uri = new URI(param);
    StreamSource source = new StreamSource(socket.getInputStream());
    XsltCompiler compiler = new Processor(true).newXsltCompiler();
    
    compiler.loadExecutablePackage(uri).load().transform();
    compiler.compilePackage(source).link().load().transform();
    compiler.loadLibraryPackage(uri).link().load().transform();
  }
```

File: java/ql/test/experimental/query-tests/security/CWE-074/options [hunk 1]
```diff
@@ -0,0 +1 @@
+//semmle-extractor-options: --javac-args -cp ${testdir}/../../../stubs/springframework-5.2.3:${testdir}/../../../stubs/Saxon-HE-9.9.1-7
```
Comment (aschackmull) [inline java/ql/test/experimental/query-tests/security/CWE-074/options:1]:
```suggestion
//semmle-extractor-options: --javac-args -cp ${testdir}/../../../../stubs/springframework-5.2.3:${testdir}/../../../stubs/Saxon-HE-9.9.1-7
```

Comment (aschackmull) [reply java/ql/test/experimental/query-tests/security/CWE-074/options:1]: Might as well reuse the existing spring stubs.

File: java/ql/test/experimental/stubs/springframework-5.2.3/org/springframework/web/bind/annotation/RequestParam.java [hunk 2]
```diff
@@ -0,0 +1,8 @@
+package org.springframework.web.bind.annotation;
```
Comment (aschackmull) [inline java/ql/test/experimental/stubs/springframework-5.2.3/org/springframework/web/bind/annotation/RequestParam.java:1]: Reusing the existing spring stubs means that this file isn't needed (it's already in `java/ql/test/stubs/springframework-5.2.3`)

File: java/ql/test/experimental/query-tests/security/CWE-074/XsltInjection.java [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+    StreamSource source = new StreamSource(socket.getInputStream());
+    TransformerFactory.newInstance().newTransformer(source).transform(null, null);
+  }
+
+  public void testStreamSourceReader(Socket socket) throws Exception {
+    StreamSource source = new StreamSource(new InputStreamReader(socket.getInputStream()));
+    TransformerFactory.newInstance().newTemplates(source).newTransformer().transform(null, null);
+  }
+
+  public void testStreamSourceInjectedParam(@RequestParam String param) throws Exception {
```
Comment (aschackmull) [inline java/ql/test/experimental/query-tests/security/CWE-074/XsltInjection.java:36]: Modelling of taint sources based on `@RequestParam` has been improved a bit, so these methods also need a `@RequestMapping` annotation in order for this to be considered a taint source (the test is currently failing due to this).

File: java/ql/test/experimental/query-tests/security/CWE-074/options (hunk 1 above)
Comment (ggolawski) [reply java/ql/test/experimental/query-tests/security/CWE-074/options:1]: I've changed it to reuse the existing stubs.

File: java/ql/test/experimental/stubs/springframework-5.2.3/org/springframework/web/bind/annotation/RequestParam.java (hunk 2 above)
Comment (ggolawski) [reply java/ql/test/experimental/stubs/springframework-5.2.3/org/springframework/web/bind/annotation/RequestParam.java:1]: Removed.

File: java/ql/test/experimental/query-tests/security/CWE-074/XsltInjection.java (hunk 3 above)
Comment (ggolawski) [reply java/ql/test/experimental/query-tests/security/CWE-074/XsltInjection.java:36]: I've added `@RequestMapping` annotation to the methods with `@RequestParam`.
However, the test was passing for me even without this annotation.

**Review by aschackmull — APPROVED** [review]

**Pull request merged**